Software/lib/GNSE_HAL/GNSE_hal.c

```diff
@@ -0,0 +1,155 @@
+/** Copyright © 2021 The Things Industries B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file GNSE_hal.c
+ *
+ * @copyright Copyright (c) 2021 The Things Industries B.V.
+ *
+ */
+
+#include "GNSE_hal.h"
+#include "GNSE_acc.h"
+#include "GNSE_bm.h"
+#include "SHTC3.h"
+#include "mxic_hc.h"
+#include "BUZZER.h"
+
+/**
+  * @brief Initialises the internal sensors for GNSE, the accelerometer and humidity/temperature sensor
+  * @note I2C interface should be initialised before this function
+  * @param none
+  * @return none
+  */
+void GNSE_Internal_Sensors_Init(void)
+{
+    GNSE_BSP_LS_Init(LOAD_SWITCH_SENSORS);
+    GNSE_BSP_LS_On(LOAD_SWITCH_SENSORS);
+    /* Delay required to let the sensors start up */
+    HAL_Delay(100);
+
+    GNSE_ACC_Init();
+    SHTC3_probe();
+}
+
+/**
+  * @brief Deinitialises the internal sensors for GNSE, the accelerometer and humidity/temperature sensor
+  * @note The I2C interface is not turned off by this function
+  * @param none
+  * @return none
+  */
+void GNSE_Internal_Sensors_DeInit(void)
+{
+    GNSE_ACC_DeInit();
+    GNSE_BSP_LS_Off(LOAD_SWITCH_SENSORS);
+    GNSE_BSP_LS_DeInit(LOAD_SWITCH_SENSORS);
+}
+
+/**
+  * @brief Initialises hardware of the flash IC
+  * @param none
+  * @return none
+  */
+void GNSE_Flash_Init(void)
+{
+    GNSE_BSP_LS_Init(LOAD_SWITCH_FLASH);
+    GNSE_BSP_LS_On(LOAD_SWITCH_FLASH);
+    MxSpi mxic = {0};
+    MxHardwareInit(&mxic);
```

Comment:
This implementation is confusing for a simple reason.
Please imagine the flow of someone initializing the flash, before implementing please provide me with the functions that the user needs to call to initialize the external flash.

Reply:
I could use MX25R16_Init. Although it also initialises the MxChip handler (and should therefor be a param). This is fine for GNSE_HAL_Flash_Init as single param, but should I also include the handler with the GNSE_HAL_Ctx_t structure?

Reply:
Yeah use it, and your issue is how to pass the handler to the application layer if used!
This can be done in multiple ways (best would be passing by reference to the API)
Something similar to what I mentioned before 
```c
MX25R16_Init(&GNSE_HAL_Flash);
```
and your API being 
```
GNSE_Flash_Init (MxChip * flash_handle)
```
This is the good way

Another (worse but easier) way is to extern like [this](https://github.com/TheThingsIndustries/generic-node-se/blob/34865ec743f5513916b99f031ccf129321dee143/Software/lib/GNSE_BSP/GNSE_bsp_serial.h#L32) and use the one singleton design pattern handler for lib and any app

Since we don't do object oriented yet (will do it in the future), both solutions are good, and in the future we can pass it through `GNSE_HAL_Ctx_t` but now it's not needed